Include/Rover/TrialView.hpp

```diff
@@ -0,0 +1,264 @@
+
+          //! Returns a reference to the Sample the iterator would point to
+          //! if it was advanced by the offset.
+          const Sample& operator [](std::ptrdiff_t offset) const;
+
+        private:
+          friend class TrialView<S>;
+          ConstIterator(const GetPtr& get_ptr, std::size_t offset);
+          const GetPtr* m_get_ptr;
+          std::size_t m_offset;
+      };
+
+      //! Creates a TrialView for a trial.
+      template<typename Trial>
+      TrialView(const Trial& t);
+
+      //! Returns an iterator to the beginning of the trial.
+      ConstIterator begin() const;
+
+      //! Returns an iterator to the end of the trial.
+      ConstIterator end() const;
+
+      //! Returns the size of the trial.
+      std::size_t size() const;
+
+      //! Returns a reference to the element at the given index.
+      const Sample& operator [](std::size_t index) const;
+
+    private:
+      std::size_t m_size;
+      GetPtr m_get_ptr;
+  };
+
+  template<typename Trial>
+  TrialView(const Trial&) -> TrialView<typename Trial::Sample>;
+
+  template<typename S>
+  TrialView<S>::ConstIterator::ConstIterator(const GetPtr& get_ptr,
+      std::size_t offset)
+    : m_get_ptr(&get_ptr),
+      m_offset(offset) {}
+
+  template<typename S>
+  typename const TrialView<S>::Sample&
+      TrialView<S>::ConstIterator::operator *() const {
+    return *(*m_get_ptr)(m_offset);
+  }
+
+  template<typename S>
+  typename const TrialView<S>::Sample*
+      TrialView<S>::ConstIterator::operator ->() const {
+    return (*m_get_ptr)(m_offset);
+  }
+
+  template<typename S>
+  typename const TrialView<S>::Sample&
+      TrialView<S>::ConstIterator::operator [](std::ptrdiff_t offset) const {
+    return *(*m_get_ptr)(m_offset + offset);
+  }
+
+  template<typename S>
+  bool TrialView<S>::ConstIterator::operator ==(ConstIterator other) const {
+    return m_offset == other.m_offset;
+  }
+
+  template<typename S>
+  bool TrialView<S>::ConstIterator::operator !=(ConstIterator other) const {
+    return m_offset != other.m_offset;
+  }
+
+  template<typename S>
+  bool TrialView<S>::ConstIterator::operator <(ConstIterator other) const {
+    return m_offset < other.m_offset;
+  }
+
+  template<typename S>
+  bool TrialView<S>::ConstIterator::operator <=(ConstIterator other) const {
+    return m_offset <= other.m_offset;
+  }
+
+  template<typename S>
+  bool TrialView<S>::ConstIterator::operator >(ConstIterator other) const {
+    return m_offset > other.m_offset;
+  }
+
+  template<typename S>
+  bool TrialView<S>::ConstIterator::operator >=(ConstIterator other) const {
+    return m_offset >= other.m_offset;
+  }
+
+  template<typename S>
+  typename TrialView<S>::ConstIterator&
+      TrialView<S>::ConstIterator::operator ++() {
+    ++m_offset;
+    return *this;
+  }
+
+  template<typename S>
+  typename TrialView<S>::ConstIterator
+      TrialView<S>::ConstIterator::operator ++(int) {
+    ++m_offset;
```

Comment:
Unfortunately post-fix increment works different than prefix increment. You will need to first make a copy of the iterator, increment ```m_offset```, then return the copy.